lib/src/test/java/io/ably/lib/test/rest/RestJWTTest.java

```diff
@@ -0,0 +1,157 @@
+	 * Base request of a JWT token with wrong credentials (RSA8g RSA8c)
+	 */
+	@Test
+	public void authjwtrequestwrongkeys() {
+		try {
+			options.token = getTokenWithParams(invalidKeys);
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+		} catch (AblyException e) {
+			assertEquals("Unexpected code from exception", 40144, e.errorInfo.code);
+			assertEquals("Unexpected statusCode from exception", 401, e.errorInfo.statusCode);
+			assertEquals("Error message not matching the expected one", "Error verifying JWT; err = invalid signature", e.errorInfo.message);
+		}
+	}
+
+	/**
+	 * Request of a JWT token that embeds and Ably token (RSC1 RSC1a RSC1c RSA3d)
+	 */
+	@Test
+	public void authjwtrequestembeddedtoken() {
+		try {
+			options.token = getTokenWithParams(mergeParams(validKeys, tokenEmbedded));
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("authjwtrequestembeddedtoken: Unexpected exception");
+		}
+	}
+
+	/**
+	 * Request of a JWT token that embeds and Ably token and is encrypted (RSC1 RSC1a RSC1c RSA3d)
+	 */
+	@Test
+	public void authjwtrequestembeddedtokenencrypted() {
+		try {
+			options.token = getTokenWithParams(mergeParams(validKeys, tokenEmbeddedAndEncrypted));
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("authjwtrequestembeddedtokenencrypted: Unexpected exception");
+		}
+	}
+
+	/**
+	 * Request of a JWT token that is returned with application/jwt content type (RSA4f, RSA8c)
+	 */
+	@Test
+	public void authjwtrequestreturntype() {
+		try {
+			options.token = getTokenWithParams(mergeParams(validKeys, jwtReturnType));
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("authjwtrequestreturntype: Unexpected exception");
+		}
+	}
+
+	/**
+	 * Request of a JWT token via authCallback (RSA8g)
+	 */
+	@Test
+	public void authjwtrequestauthcallback() {
+		try {
+			restJWTRequester = new AblyRest(createOptions(testVars.keys[0].keyStr));
+			TokenCallback authCallback = new TokenCallback() {
+				@Override
+				public Object getTokenRequest(TokenParams params) throws AblyException {
+					return restJWTRequester.auth.createTokenRequest(params, null);
+				}
+			};
+			ClientOptions optionsWithCallback = new ClientOptions();
+			optionsWithCallback.authCallback = authCallback;
+			optionsWithCallback.environment = "sandbox";
+			AblyRest client = new AblyRest(optionsWithCallback);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
```

Comment:
I think we need an assertion that proves that the callback was called.

Reply:
done